sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1134,28 +1119,136 @@ _pi_queue::_pi_queue(std::vector<ze_command_queue_handle_t> &ComputeQueues,
   CopyCommandBatch.QueueBatchSize = ZeCommandListBatchCopyConfig.startSize();
 }
 
+// Perform any necessary cleanup after an event has been signalled.
+// This currently makes sure to release any kernel that may have been used by
+// the event, updates the last command event in the queue and cleanes up all dep
+// events of of the event.
+// The caller must not lock any mutexes.
+static pi_result cleanup(pi_event Event) {
+  pi_kernel AssociatedKernel = nullptr;
+  // List of dependent events.
+  std::list<pi_event> EventsToBeReleased;
+  pi_queue AssociatedQueue = nullptr;
+  {
+    std::scoped_lock EventLock(Event->Mutex);
+    // Exit early of event was already cleanedup.
+    if (Event->CleanedUp)
+      return PI_SUCCESS;
+
+    AssociatedQueue = Event->Queue;
+
+    // Remember the kernel associated with this event if there is one. We are
+    // going to release it later.
+    if (Event->CommandType == PI_COMMAND_TYPE_NDRANGE_KERNEL &&
+        Event->CommandData) {
+      AssociatedKernel = pi_cast<pi_kernel>(Event->CommandData);
+      Event->CommandData = nullptr;
+    }
+
+    // Make a list of all the dependent events that must have signalled
+    // because this event was dependent on them.
+    Event->WaitList.collectEventsForReleaseAndDestroyPiZeEventList(
+        EventsToBeReleased);
+
+    Event->CleanedUp = true;
+  }
+
+  // We've reset event data members above, now cleanup resources.
+  if (AssociatedKernel)
+    PI_CALL(piKernelRelease(AssociatedKernel));
+
+  if (AssociatedQueue) {
+    // Lock automatically releases when this goes out of scope.
+    std::scoped_lock Lock(AssociatedQueue->Mutex);
+
+    // If this event was the LastCommandEvent in the queue, being used
+    // to make sure that commands were executed in-order, remove this.
+    // If we don't do this, the event can get released and freed leaving
+    // a dangling pointer to this event.  It could also cause unneeded
+    // already finished events to show up in the wait list.
+    if (AssociatedQueue->LastCommandEvent == Event) {
+      AssociatedQueue->LastCommandEvent = nullptr;
```

Comment:
If possible, it might be better to wrap `LastCommandEvent` into `std::atomic` and do `compare_exchange_strong` without mutex lock.